File: src/main/java/io/Cbuild/gnu_make_functions/compiler/overwrite/compile_and.java
package io.Cbuild.gnu_make_functions.compiler.overwrite;

import io.Cbuild.cBuildIR;
import io.Cbuild.cbuildException;
import io.Cbuild.cbuildParser;
import io.Cbuild.gnu_make_functions.compiler.compile_fn;
import io.Cbuild.gnu_make_functions.compiler.util;
import io.Cbuild.gnu_make_functions.type;

public class compile_and extends compile_fn {

    @Override
    public cBuildIR.FunctionIR compile(cbuildParser.FunctionContext ctx, type.MakeFunction function) {
        cBuildIR.FunctionIR ir = super.compile(ctx, function);
        util.cleanWS(ir);

        if (ir.args.isEmpty()) {
            throw new cbuildException(
                    cbuildException.ErrorType.SEMANTIC,
                    "and: expected at least 1 argument: cond1[, cond2...]",
                    ctx.start.getLine(),
                    ctx.start.getCharPositionInLine() + 1
            );
        }

        return ir;
    }
}

File: src/gnu-make-functions/compiler/overwrite/compile_and.ts
import { FunctionIR } from "@compiler/ir.js";
import { FunctionContext } from "@parser/cbuildParser.js";
import { cbuildException, ErrorType } from "@src/cbuild-exception.js";
import type { MakeFunction } from "@gnu-make-functions/type.js";
import { compile_fn } from "@gnu-make-functions/compiler/compile_fn.js";
import { util } from "@gnu-make-functions/compiler/util.js";

export class compile_and extends compile_fn {
  public override compile(
    ctx: FunctionContext,
    func: MakeFunction,
  ): FunctionIR {
    const ir = super.compile(ctx, func);
    util.cleanWS(ir);

    if (ir.args.length === 0) {
      throw new cbuildException(
        ErrorType.SEMANTIC,
        "and: expected at least 1 argument: cond1[, cond2...]",
        ctx.start?.line || 0,
        (ctx.start?.column || 0) + 1,
      );
    }

    return ir;
  }
}
